README.md
# Program Translate




### ความสัมพันธ์และที่มาของตัวโปรแกรม
**Program Translate** เป็นโปรเจคกลุ่มนักศึกษามหาวิทยาลัยเชียงใหม่ ชั้นปีที่ 1 ปีการศึกษา 2561 ทำขึ้นในรายวิชา 204113 [Principles of Computing] โดยตัวโปรแกรมของเรานั้นเป็นโปรแกรมที่จะเอาคำศัพท์มาจาก หนังสือ "พจนานุกรมคอมพิวเตอร์และอินเตอร์เน็ต" เพื่อแปลคำศัพท์จากภาษาอังกฤษเป็นภาษาไทยตามหนังสือที่กล่าวมาข้างต้น

### การติดตั้งโปรแกรม :file_folder:
1. โหลดโปรแกรมไฟล์จากลิงค์ :link:[Program Translate](https://drive.google.com/file/d/1OiG_fLLs4WIyb1eVpexjgxlRPBmPcmqj/view?usp=sharing)
1. แตกไฟล์ **Program Translate.winrar** และเราจะได้โฟลเดอร์ที่เป็นโฟลเดอร์ของตัวโปรแกรมมาพร้อม modual ต่างๆที่ทำให้โปรแกรมสามารถใช้งานได้<p>
![1](https://user-images.githubusercontent.com/49282113/56789530-495a1180-682c-11e9-934b-046a22554980.png)<p>

1. กดที่ไฟล์ **'launch.bat'** เพื่อเปิดตัวโปรแกรมใหสามารถทำงานได้<p>
![3](https://user-images.githubusercontent.com/49282113/56789696-b372b680-682c-11e9-91af-43479965da93.png)


### การใช้งาน :computer:
ตัวโปรแกรมจะมีช่องให้กรอกข้อความภาษาอังกฤษ ที่เราต้องการรู้ความหมายกดปุ่ม Translate บนตัวโปรแกรม/ กดปุ่ม Enter บนคีบอร์ด โปรแกรมจะแสดงความหมายของคำออกมาในคำอธิบายด้านล่าง<p>
1. ช่องพิมพ์คำศัพท์ทางคอมพิวเตอร์ที่เป็นภาษาอังกฤษ<p>
1. ช่องอธิบายคำแปล<p>
<img src="https://user-images.githubusercontent.com/49282113/56789734-d3a27580-682c-11e9-8f09-e4790b78bbe9.png" width="400">

 


### ฟังก์ชันเสริม
> **Click!! >> [File]** 

![file](https://user-images.githubusercontent.com/49282113/56809034-1aa85f00-685d-11e9-80fa-9f74722c5759.png)

* [Top 5 Serach]  : หน้าต่างที่จะบอกถึงการค้นหาคำศัพท์ที่มากที่สุด 5 อันดับแรกพร้อมจำนวนครั้งที่ค้นหา <p>
![top5](https://user-images.githubusercontent.com/49282113/56809258-a621f000-685d-11e9-857d-26abb679b44f.png)

* [History]  : ปุ่มที่จะแสดงประวัติการค้นหา 10 อันดับล่าสุด และแสดงวัน/เดือน/ปี, เวลาที่ทำการค้นหา <p>
<img src="https://user-images.githubusercontent.com/49282113/56809647-8e973700-685e-11e9-94f8-09d55a5c6bca.png" width="400">

* [Exit] : ปิดระบบโปรแกรม
<p>
<p>
 
> **Click!! >> [Help]**

![Help](https://user-images.githubusercontent.com/49282113/56810387-6577a600-6860-11e9-96fc-c24579b1b505.png)

* [Developer] : รายชื่อสมาชิกผู้จัดทำโปรแกรม<p>
<img src="https://user-images.githubusercontent.com/49282113/56810159-c783db80-685f-11e9-9473-eacb09353460.png" width="400">

* [About..] : หน้าแสดงที่มาของตัวโปรแกรม ทั้งอ้างอิงเอกสาร และโปรแกรมที่ใช้งาน
<img src="https://user-images.githubusercontent.com/49282113/56810819-6ceb7f00-6861-11e9-92af-bdef70fbfc87.png" width="400">




### รายชื่อสมาชิกกลุ่ม
:trollface: น.ส.วชิรดา ท้าวนอก 600510575
<p>
:trollface: น.ส.ชนกานต์ เอกอนันต์กุล 610510648
<p>
:trollface: น.ส.นนทรพร จันทร์มณีวงศ์ 610510654
<p>
:trollface: น.ส.จุฬาลักษณ์ สุจันทร์ 610510801
<p>
:trollface: นายธเนศ สิงห์ลอ 610510803
<p>
:trollface: นายวสันต์ แพทย์รัตน์ 610510809


### Poster :camera:
<p align="center">
<img src="https://user-images.githubusercontent.com/49282113/56810899-9dcbb400-6861-11e9-92af-1bb7414554ec.png" width="600">



Program Translate/history.txt
file;2
firewall;2


Program Translate/main.py
#!/usr/bin/env python3
# Module check file of word

# Main of Program
import listfile as fileloc
import module_addword as modAddword
import module_checkfile as modCheckfile
import module_find as modFind
import module_history as modHist

def main():
    print("Start Computer Dictionary")
    print("-------------------------------------------------------")
    print("Staus: checking file")
    st_modCheckfile = modCheckfile.main()
    if st_modCheckfile:
        print("Status: check file Completed")
    else:
        print("Status: check file Error")
    print("Program ready to use")
    print("-------------------------------------------------------")
    while True:
        print("Key for control program")
        print("a = add new word")
        print("f = find word")
        print("h = show history")
        print("ht = show top history")
        print("esc = escape program")
        key_input = input("key: ")
        print("-------------------------------------------------------")
        hist = modHist.main()
        if key_input == "a":
            modAddword.main()
        elif key_input == "f":
            modFind.main()
        elif key_input == "h":
            hist.print()
        elif key_input == "ht":
            hist.top()
        elif key_input == "esc":
            break
        else:
            continue
        print("-------------------------------------------------------")

if __name__ == '__main__':
    main()

Program Translate/module_find.py
#!/usr/bin/env python3
# Module find word

import listfile as fileloc
import module_history as modHist

def main(word):
    char_word = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
                'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
    inword_lst = list(word.lower())
    if word == "":
        return None
    if inword_lst[0] in char_word:
        word = word.lower()
        a = Findword(word)
        return a.find()
    else:
        return None

class Findword(object):
    def __init__(self, word):
        self.inword = word
        self.id = None
        self.word = None
        self.wmean = None
        self.group = None
    def find(self):
        openfile = open(fileloc.FOLDER_WORD+self.inword[0]+".txt", "r",encoding="utf8")
        i = 0;
        for line in openfile:
            if i == 0:
                i += 1
                continue
            line = line.split(";")
            line[2] = line[2].replace("\n", "")
            list_word = list(line[1])
            list_word[0] = list_word[0].upper()
            full_word = "".join(list_word)
            self.id = line[0]
            self.word = line[1].lower()
            self.wmean = line[2]
            #self.group = line[3]
            if self.word == self.inword:    # เงื่อนไขเมื่อ ข้อมูลในไฟล์ = ข้อมูลเข้า
                hist = modHist.main()
                hist.add(self.word)
                hist.addhist(self.inword)
                return self.wmean
        openfile.close()
        return None

if __name__ == '__main__':
    main()

Program Translate/module_addword.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Module add word for new word

import listfile as fileloc

def main():
    w_word = input("Word: ")
    w_mean = input("Mean: ")
    w_group = int(input("Group: "))
    add = Addword(w_word, w_mean, w_group)
    add.add()

class Addword(object):
    def __init__(self, word, word_mean, group):
        self.id = None
        self.word = word
        self.wmean = word_mean
        self.group = group
    def add(self):
        split_first_word = list(self.word)
        self.checkID(split_first_word)
        openfile = open(fileloc.FOLDER_WORD+split_first_word[0].upper()+".txt", "a")
        openfile.write("%d;%s;%s;%s\n" % (self.id, self.word.lower(), self.wmean, self.group))
        openfile.close()
        print("Add word %s: Completed" % self.word)
    def checkID(self, split_word):
        openfile = open(fileloc.FOLDER_WORD+split_word[0]+".txt", "r")
        for line in openfile:
            self.id = int(line) + 1
            break
        with open(fileloc.FOLDER_WORD+split_word[0]+".txt") as f:
            lines = f.readlines()
        lines[0] = str(self.id)+"\n"
        with open(fileloc.FOLDER_WORD+split_word[0]+".txt", "w") as f:
            f.writelines(lines)
        print(self.id)
        openfile.close()

if __name__ == '__main__':
    main()

Program Translate/module_history.py
#!/usr/bin/env python3
# Module check file of word

import listfile as fileloc
from operator import itemgetter
import datetime

def main():
    return History()

class History(object):
    def __init__(self):
        self.word = None
        self.rank = None
    def add(self, word):
        openfile = open("history.txt", "r+", encoding="utf8")
        for line in openfile:
            list_line = line.split(";")
            if list_line[0] == word:
                pre_edit = ("%s;%s" % (list_line[0], list_line[1]))
                post_edit = ("%s;%s\n" % (list_line[0], (int(list_line[1])+1)))
                change = open("history.txt", encoding='utf8').read().replace(pre_edit, post_edit)
                f = open("history.txt", 'w', encoding="utf8")
                f.write(change)
                f.close()
                openfile.close()
                return
        self.word = word
        self.rank = "1"
        openfile.write("%s;%s\n" % (self.word.lower(), self.rank))
        openfile.close()
    def top(self):
        rank_list = []
        openfile = open("history.txt", "r")
        for line in openfile:
            list_line = line.split(";")
            try:
                score = list_line[1].replace("\n", "")
            except IndexError:
                break
            rank_list.append([list_line[0], score])
        openfile.close()
        for i in rank_list:
            i[1] = int(i[1])
        rank_list = sorted(rank_list, key=itemgetter(1), reverse=True)
        print(rank_list)
        return rank_list
#-------------------------------------------------------------------------------------#
    def addhist(self, word):    
        list_openfile = []
        openfile = open("historyword.txt", "r")
        for line in openfile:
            list_openfile.append(line.split(";"))
        openfile.close()
        for i in range(len(list_openfile)):
            if list_openfile[i] == "":
                list_openfile.pop(i)
            if word == list_openfile[i][0]:
                list_openfile[i][1] = str(datetime.datetime.now()).strip('datetime.datetime')+"\n"
                self.addhist_(list_openfile)
                return
        list_openfile.append([word, (str(datetime.datetime.now()).strip('datetime.datetime')+"\n")])
        self.addhist_(list_openfile)
    def addhist_(self, list_openfile):
        clearfile = open("historyword.txt", "w+")
        clearfile.close()
        openfile = open("historyword.txt", "a+")
        for i in list_openfile:
            openfile.write(i[0]+";"+i[1])
        openfile.close()

#-------------------------------------------------------------------------------------#
    def check(self, word):
        openfile = open("historyword.txt", "r")
        for line in openfile:
            if line == word+"\n":
                return True
            else:
                return False
        openfile.close()
    def printhist(self):
        list_openfile = []
        openfile = open("historyword.txt", "r")
        for line in openfile:
            list_openfile.append(line.split(";"))
        openfile.close()
        list_openfile = sorted(list_openfile, key=itemgetter(1), reverse=True)
        return list_openfile
    def print(self):
        openfile = open("history.txt", "r")
        for line in openfile:
            linenew = line.split(";")
            print(linenew[0])
        openfile.close()
    def clear(self):
        openfile = open("history.txt", "w+")
        openfile.close()
        openfile = open("historyword.txt", "w+")
        openfile.close()

if __name__ == '__main__':
    main()

Program Translate/module_checkfile.py
#!/usr/bin/env python3
# Module check file of word

import listfile as fileloc

def main():
    print("Start operation: Checkfile")
    checkfile = Checkfile()
    checkfile.checkFile()

class Checkfile(object):
    def __init__(self):
        self.filelist = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m",
                        "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"]
    def checkFile(self):
        for i in self.filelist:
            openfile = open(fileloc.FOLDER_WORD+i+".txt", "r+",encoding="utf8")
            for line in openfile:
                if line == "":
                    openfile.write("0\n")
                break
            openfile.close()
            print("Check file %s: Founded" % i)

if __name__ == '__main__':
    main()

Program Translate/Main_program.py
from tkinter import *
from tkinter import messagebox
import listfile as fileloc
import module_addword as modAddword
import module_checkfile as modCheckfile
import module_find as modFind
import module_history as modHist

main = Tk()

main.bind("<Escape>", exit)
main.geometry("700x700")
main.resizable(width = False, height = False)


main.title('Program Translate')

background_image = PhotoImage(file='landscape2.png')
background_label = Label(main, image=background_image)
background_label.place(relwidth=1, relheight=1)

#--------------------------------------------------------------
def Pop_result():
  val = entry.get()
  result = modFind.main(val)
  root = Tk()
  root.title('คำแปล')
  S = Scrollbar(root)
  T = Text(root, height=18, width=50,font=("Courier", 16))
  S.pack(side=RIGHT, fill=Y)
  T.pack(side=LEFT, fill=Y)
  S.config(command=T.yview)
  T.config(yscrollcommand=S.set)
  quote = result
  T.insert(END, quote)
  T.config(state=DISABLED)
  mainloop()
#--------------------------------------------------------------
def Pop_Rank():
  hist = modHist.main()
  list_hist = hist.top()
  hist_word = ""
  j = 1
  main = Tk()
  main.geometry('500x225')
  main.resizable(width = False, height = False)
  main.title('Top 5 Search')

  frame = Frame(main)
  frame.place(relheight=1, relwidth=1)

  for i in list_hist:
    if j == 6:
      break
    hist_word += "%d. %s %s ครั้ง\n" % (j, i[0], i[1])
    j += 1

  msg1 = Message(frame,font=("Perpetua", 20),text = hist_word,width = 300)#พื้นหลัง
  msg1.place(relx=0.25,rely=0, relwidth=1, relheight=1, anchor='n')#พื้นหลัง
#--------------------------------------------------------------
def ask_user():
  if messagebox.askretrycancel('Check','Not found, Please try again',) == True:
    entry.delete(0, END)
  else:
    pass
#---------------------------------------------------------
def to_History():
  hist = modHist.main()
  list_hist = hist.printhist()
  hist_word = ""
  j = 1
  main = Tk()
  main.geometry('700x700')
  main.resizable(width = False, height = False)
  main.title('History')

  frame = Frame(main)
  frame.place(relheight=1, relwidth=1)

  label = Label(main,text ='History',font = ("Perpetua", 20), fg = 'blue')
  label.place(relx=0.5, rely=0.0000000000000000000000000001, relwidth=0.75, relheight=0.1,anchor='n')

  for i in list_hist:
    if j == 11:
      break
    hist_word += "%d. %s %s\n" % (j, i[0], i[1])
    j += 1

  msg = Message(frame,font=("Perpetua", 12),text =hist_word,width = 1000)#พื้นหลัง
  msg.place(relx=0.5,rely=0.01, relwidth=10, relheight=1, anchor='n')#พื้นหลัง

  button = Button(frame, text="Clear History", command = (clear_hist),bg = 'gray' ,fg = 'white')
  button.place(relx=0.5, rely=0.87, relwidth=0.25, relheight=0.1,anchor='n')
#---------------------------------------------------------
def clear_hist():
  messagebox.showinfo("Clear", "Done!")
  hist = modHist.main()
  hist.clear()
#---------------------------------------------------------
def about():
  main = Tk()
  main.geometry('550x300')
  main.resizable(width = False, height = False)
  main.title('เอกสารอ้างอิง & โปรแกรม')

  label = Label(main,text ='เอกสารที่ใช้ : พจนานุกรมคอมพิวเตอร์และอินเตอร์เน็ต \n (ผู้เขียน ทีมบรรณาธิการหนังสือคอมพิวเตอร์) \n ------------------------------------------------- \n โปรแกรมที่ใช้ : Python (version 3.7),กล้องถ่ายรูป,Google Sheets \n Adobe Photoshop CC,Adobe Photoshop C6,Sublime Text \n Microsoft Powerpoint,Microsoft Word,Github' ,font = ("Perpetua", 15), fg = 'blue')
  label.place(relx=0.5, relwidth=10, relheight=1, anchor='n')
  
#---------------------------------------------------------
def develop():
  main = Tk()
  main.geometry('650x450')
  main.resizable(width = False, height = False)
  main.title('Group : She sell seashell by the seashore')

  frame = Frame(main)
  frame.place(relheight=1, relwidth=1)

  label = Label(main,text ='ผู้จัดทำ',font = ("Perpetua", 30), fg = 'blue')
  label.place(relx=0.5, rely=0.09, relwidth=0.75, relheight=0.1,anchor='n')

  msg = Message(frame,font=("Perpetua", 20),text = 'น.ส.วชิรดา ท้าวนอก  600510575\nน.ส.ชนกานต์ เอกอนันต์กุล  610510648\nน.ส.นนทรพร จันทร์มณีวงศ์  610510654\nน.ส.จุฬาลักษณ์ สุจันทร์  610510801\nนายธเนศ สิงห์ลอ  610510803\nนายวสันต์ แพทย์รัตน์  610510809',width = 500)#พื้นหลัง
  msg.place(relx=0.5,rely=0.01, relwidth=1, relheight=1, anchor='n')#พื้นหลัง

#---------------------------------------------------------Menu Bar
menubar = Menu(main)
filemenu = Menu(menubar, tearoff=0)
filemenu.add_command(label="Top 5 Search", command=Pop_Rank)
filemenu.add_command(label="History", command=to_History)
filemenu.add_separator()
filemenu.add_command(label="Exit", command=main.quit)
menubar.add_cascade(label="File", menu=filemenu)
helpmenu = Menu(menubar, tearoff=0)
helpmenu.add_command(label="Developer", command=develop)
helpmenu.add_command(label="About...", command=about)
menubar.add_cascade(label="Help", menu=helpmenu)

main.config(menu=menubar)

#--------------------------------------------------------- เฟรม Program Translate
top_frame = Frame(main, bg='white', bd=10)
top_frame.place(relx=0.28, rely=0.05, relwidth=0.45, relheight=0.125)

label = Label(top_frame,text ='Program Translate',font = ("Perpetua", 20), bg='#BFE9FF')
label.place(relwidth=1, relheight=1)

#---------------------------------------------------------เฟรม กรอกข้อความ
frame = Frame(main, bg='#80c1ff', bd=5)
frame.place(relx=0.5, rely=0.2, relwidth=0.75, relheight=0.1,anchor='n')

entry = Entry(frame, font=40,justify='center')
entry.place(relwidth=1, relheight=1)

#---------------------------------------------------------เฟรม ปุ่ม
frame = Frame(main, bg='#80c1ff', bd=5)
frame.place(relx=0.5, rely=0.32, relwidth=0.25, relheight=0.1,anchor='n')
#---------------------------------------------------------

def click_translate():
  msg1 = Message(lower_frame,font=("Perpetua", 20),bd = 3)#พื้นหลัง
  msg1.place(relx=0.5, rely=0, relwidth=1, relheight=1, anchor='n')#พื้นหลัง
  msg2 = Message(lower_frame,text = 'คำอธิบาย',font=("Cordia New", 18),bd = 3,width = 100000) #คำแปล
  msg2.place(relx=0.4)
  val = entry.get()
  result = modFind.main(val)

  if result == None:
    ask_user()
  elif len(result) <= 200:
    msg = Message(lower_frame,text =result,font=("Cordia New", 16),width=420)  #ยาว 30 คำ 6 บรรทัด
    msg.place(relx=0.05, rely=0.1341)

  elif len(result) >= 201:
    Pop_result()

#------------------------------------------------------------------ Enter to trans
def enter(obj):
  msg1 = Message(lower_frame,font=("Perpetua", 20),bd = 3)#พื้นหลัง
  msg1.place(relx=0.5, rely=0, relwidth=1, relheight=1, anchor='n')#พื้นหลัง
  msg2 = Message(lower_frame,text = 'คำอธิบาย',font=("Cordia New", 18),bd = 3,width = 100000) #คำแปล
  msg2.place(relx=0.4)#คำแปล

  val = entry.get()
  result = modFind.main(val)
  if result == None:
    ask_user()
  elif len(result) <= 200:
    msg = Message(lower_frame,text =result,font=("Cordia New", 16),width=420)  #ยาว 30 คำ 6 บรรทัด
    msg.place(relx=0.05, rely=0.1341)

  elif len(result) >= 201:
    Pop_result()

button = Button(frame, text="Translate", font=40, command = click_translate)
button.place(relx=0.5, rely=0.009, relwidth=1, relheight=1,anchor='n')
main.bind('<Return>', enter)
#------------------------------------------------------------------

lower_frame = Frame(main, bg='#80c1ff', bd=10)
lower_frame.place(relx=0.5, rely=0.45, relwidth=0.75, relheight=0.5, anchor='n')
  


msg1 = Message(lower_frame,font=("Cordia New", 20),bd = 3)#พื้นหลัง
msg1.place(relx=0.5, rely=0, relwidth=1, relheight=1, anchor='n')#พื้นหลัง

msg2 = Message(lower_frame,text = 'คำอธิบาย',font=("Cordia New", 18),bd = 3,width = 100000) #คำแปล
msg2.place(relx=0.4)#คำแปล



main.mainloop()
